// @ionic/cli-utils/lib/serve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_framework_1 = require("@ionic/cli-framework");
const fs_1 = require("@ionic/cli-framework/utils/fs");
const network_1 = require("@ionic/cli-framework/utils/network");
const process_1 = require("@ionic/cli-framework/utils/process");
const string_1 = require("@ionic/cli-framework/utils/string");
const chalk_1 = require("chalk");
const Debug = require("debug");
const events_1 = require("events");
const lodash = require("lodash");
const os = require("os");
const path = require("path");
const split2 = require("split2");
const through2 = require("through2");
const constants_1 = require("../constants");
const guards_1 = require("../guards");
const errors_1 = require("./errors");
const events_2 = require("./events");
const hooks_1 = require("./hooks");
const debug = Debug('ionic:cli-utils:lib:serve');
exports.DEFAULT_DEV_LOGGER_PORT = 53703;
exports.DEFAULT_LIVERELOAD_PORT = 35729;
exports.DEFAULT_SERVER_PORT = 8100;
exports.DEFAULT_LAB_PORT = 8200;
exports.DEFAULT_DEVAPP_COMM_PORT = 53233;
exports.BIND_ALL_ADDRESS = '0.0.0.0';
exports.LOCAL_ADDRESSES = ['localhost', '127.0.0.1'];
exports.BROWSERS = ['safari', 'firefox', process.platform === 'win32' ? 'chrome' : (process.platform === 'darwin' ? 'google chrome' : 'google-chrome')];
// npm script name
exports.SERVE_SCRIPT = 'ionic:serve';
exports.COMMON_SERVE_COMMAND_OPTIONS = [
    {
        name: 'address',
        summary: 'Use specific address for the dev server',
        default: exports.BIND_ALL_ADDRESS,
        groups: [cli_framework_1.OptionGroup.Advanced],
    },
    {
        name: 'port',
        summary: 'Use specific port for HTTP',
        default: exports.DEFAULT_SERVER_PORT.toString(),
        aliases: ['p'],
        groups: [cli_framework_1.OptionGroup.Advanced],
    },
    {
        name: 'livereload',
        summary: 'Do not spin up dev server--just serve files',
        type: Boolean,
        default: true,
    },
    {
        name: 'engine',
        summary: `Target engine (e.g. ${['browser', 'cordova'].map(e => chalk_1.default.green(e)).join(', ')})`,
        groups: [cli_framework_1.OptionGroup.Advanced],
    },
    {
        name: 'platform',
        summary: `Target platform on chosen engine (e.g. ${['ios', 'android'].map(e => chalk_1.default.green(e)).join(', ')})`,
        groups: [cli_framework_1.OptionGroup.Advanced],
    },
];
class ServeRunner extends events_1.EventEmitter {
    constructor() {
        super(...arguments);
        this.devAppConnectionMade = false;
    }
    createOptionsFromCommandLine(inputs, options) {
        const separatedArgs = options['--'];
        if (options['local']) {
            options['address'] = 'localhost';
            options['devapp'] = false;
        }
        const engine = this.determineEngineFromCommandLine(options);
        const address = options['address'] ? String(options['address']) : exports.BIND_ALL_ADDRESS;
        const labPort = string_1.str2num(options['lab-port'], exports.DEFAULT_LAB_PORT);
        const port = string_1.str2num(options['port'], exports.DEFAULT_SERVER_PORT);
        return {
            '--': separatedArgs ? separatedArgs : [],
            address,
            browser: options['browser'] ? String(options['browser']) : undefined,
            browserOption: options['browseroption'] ? String(options['browseroption']) : undefined,
            devapp: engine === 'browser' && (typeof options['devapp'] === 'undefined' || options['devapp']) ? true : false,
            engine,
            externalAddressRequired: options['externalAddressRequired'] ? true : false,
            lab: options['lab'] ? true : false,
            labHost: options['lab-host'] ? String(options['lab-host']) : 'localhost',
            labPort,
            livereload: typeof options['livereload'] === 'boolean' ? Boolean(options['livereload']) : true,
            open: options['open'] ? true : false,
            platform: options['platform'] ? String(options['platform']) : undefined,
            port,
            proxy: typeof options['proxy'] === 'boolean' ? Boolean(options['proxy']) : true,
            ssl: false,
            project: options['project'] ? String(options['project']) : undefined,
        };
    }
    determineEngineFromCommandLine(options) {
        if (options['engine']) {
            return String(options['engine']);
        }
        if (options['cordova']) {
            return 'cordova';
        }
        return 'browser';
    }
    displayDevAppMessage(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pkg = yield this.e.project.requirePackageJson();
            // If this is regular `ionic serve`, we warn the dev about unsupported
            // plugins in the devapp.
            if (options.devapp && guards_1.isCordovaPackageJson(pkg)) {
                const plugins = yield this.getSupportedDevAppPlugins();
                const packageCordovaPlugins = Object.keys(pkg.cordova.plugins);
                const packageCordovaPluginsDiff = packageCordovaPlugins.filter(p => !plugins.has(p));
                if (packageCordovaPluginsDiff.length > 0) {
                    this.e.log.warn('Detected unsupported Cordova plugins with Ionic DevApp:\n' +
                        `${packageCordovaPluginsDiff.map(p => `- ${chalk_1.default.bold(p)}`).join('\n')}\n\n` +
                        `App may not function as expected in Ionic DevApp.`);
                }
            }
        });
    }
    beforeServe(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const hook = new ServeBeforeHook(this.e);
            try {
                yield hook.run({ name: hook.name, serve: options });
            }
            catch (e) {
                if (e instanceof cli_framework_1.BaseError) {
                    throw new errors_1.FatalException(e.message);
                }
                throw e;
            }
        });
    }
    run(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.beforeServe(options);
            const details = yield this.serveProject(options);
            const devAppDetails = yield this.gatherDevAppDetails(options, details);
            const labDetails = options.lab ? yield this.runLab(options, details) : undefined;
            if (devAppDetails) {
                const devAppName = yield this.publishDevApp(options, devAppDetails);
                devAppDetails.channel = devAppName;
            }
            const localAddress = `${details.protocol}://localhost:${details.port}`;
            const fmtExternalAddress = (address) => `${details.protocol}://${address}:${details.port}`;
            const labAddress = labDetails ? `${labDetails.protocol}://${labDetails.address}:${labDetails.port}` : undefined;
            this.e.log.nl();
            this.e.log.info(`Development server running!` +
                (labAddress ? `\nLab: ${chalk_1.default.bold(labAddress)}` : '') +
                `\nLocal: ${chalk_1.default.bold(localAddress)}` +
                (details.externalNetworkInterfaces.length > 0 ? `\nExternal: ${details.externalNetworkInterfaces.map(v => chalk_1.default.bold(fmtExternalAddress(v.address))).join(', ')}` : '') +
                (devAppDetails && devAppDetails.channel ? `\nDevApp: ${chalk_1.default.bold(devAppDetails.channel)} on ${chalk_1.default.bold(os.hostname())}` : '') +
                `\n\n${chalk_1.default.yellow('Use Ctrl+C to quit this process')}`);
            this.e.log.nl();
            if (options.open) {
                const openAddress = labAddress ? labAddress : localAddress;
                const openURL = this.modifyOpenURL(openAddress, options);
                const opn = yield Promise.resolve().then(() => require('opn'));
                yield opn(openURL, { app: options.browser, wait: false });
                this.e.log.info(`Browser window opened to ${chalk_1.default.bold(openURL)}!`);
                this.e.log.nl();
            }
            events_2.emit('serve:ready', details);
            this.scheduleAfterServe(options, details);
            return details;
        });
    }
    afterServe(options, details) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const hook = new ServeAfterHook(this.e);
            try {
                yield hook.run({ name: hook.name, serve: lodash.assign({}, options, details) });
            }
            catch (e) {
                if (e instanceof cli_framework_1.BaseError) {
                    throw new errors_1.FatalException(e.message);
                }
                throw e;
            }
        });
    }
    scheduleAfterServe(options, details) {
        process_1.onBeforeExit(() => tslib_1.__awaiter(this, void 0, void 0, function* () { return this.afterServe(options, details); }));
    }
    gatherDevAppDetails(options, details) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (options.devapp) {
                const { computeBroadcastAddress } = yield Promise.resolve().then(() => require('./devapp'));
                // TODO: There is no accurate/reliable/realistic way to identify a WiFi
                // network uniquely in NodeJS. But this is where we could detect new
                // networks and prompt the dev if they want to "trust" it (allow binding to
                // 0.0.0.0 and broadcasting).
                const interfaces = network_1.getExternalIPv4Interfaces()
                    .map(i => (Object.assign({}, i, { broadcast: computeBroadcastAddress(i.address, i.netmask) })));
                const { port } = details;
                // the comm server always binds to 0.0.0.0 to target every possible interface
                const commPort = yield network_1.findClosestOpenPort(exports.DEFAULT_DEVAPP_COMM_PORT, '0.0.0.0');
                return { port, commPort, interfaces };
            }
        });
    }
    publishDevApp(options, details) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (options.devapp) {
                const { createCommServer, createPublisher } = yield Promise.resolve().then(() => require('./devapp'));
                const publisher = yield createPublisher(this.e.project.config.get('name'), details.port, details.commPort);
                const comm = yield createCommServer(publisher.id, details.commPort);
                publisher.interfaces = details.interfaces;
                comm.on('error', (err) => {
                    debug(`Error in DevApp service: ${String(err.stack ? err.stack : err)}`);
                });
                comm.on('connect', (data) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    if (!this.devAppConnectionMade) {
                        this.devAppConnectionMade = true;
                        yield this.displayDevAppMessage(options);
                    }
                    this.e.log.info(`DevApp connection established from ${chalk_1.default.bold(data.email)}`);
                }));
                publisher.on('error', (err) => {
                    debug(`Error in DevApp service: ${String(err.stack ? err.stack : err)}`);
                });
                try {
                    yield comm.start();
                }
                catch (e) {
                    this.e.log.error(`Could not create DevApp comm server: ${String(e.stack ? e.stack : e)}`);
                }
                try {
                    yield publisher.start();
                }
                catch (e) {
                    this.e.log.error(`Could not publish DevApp service: ${String(e.stack ? e.stack : e)}`);
                }
                return publisher.name;
            }
        });
    }
    getSupportedDevAppPlugins() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const p = path.resolve(constants_1.ASSETS_DIRECTORY, 'devapp', 'plugins.json');
            const plugins = yield fs_1.fsReadJsonFile(p);
            if (!Array.isArray(plugins)) {
                throw new Error(`Cannot read ${p} file of supported plugins.`);
            }
            // This one is common, and hopefully obvious enough that the devapp doesn't
            // use any splash screen but its own, so we mark it as "supported".
            plugins.push('cordova-plugin-splashscreen');
            return new Set(plugins);
        });
    }
    runLab(options, details) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const labDetails = {
                protocol: options.ssl ? 'https' : 'http',
                address: options.labHost,
                port: yield network_1.findClosestOpenPort(options.labPort, options.labHost),
            };
            if (options.ssl) {
                const sslConfig = this.e.project.config.get('ssl');
                if (sslConfig && sslConfig.key && sslConfig.cert) {
                    labDetails.ssl = { key: sslConfig.key, cert: sslConfig.cert };
                }
                else {
                    throw new errors_1.FatalException(`Both ${chalk_1.default.green('ssl.key')} and ${chalk_1.default.green('ssl.cert')} config entries must be set.\n` +
                        `See ${chalk_1.default.green('ionic serve --help')} for details on using your own SSL key and certificate for Ionic Lab and the dev server.`);
                }
            }
            const appUrl = `${details.protocol}://localhost:${details.port}`;
            try {
                yield this.runLabServer(appUrl, labDetails);
            }
            catch (e) {
                if (e.code === 'ENOENT') {
                    const pkg = '@ionic/lab';
                    const requiredMsg = `This package is required for Ionic Lab. For more details, please see the CHANGELOG: ${chalk_1.default.bold('https://github.com/ionic-team/ionic-cli/blob/master/packages/ionic/CHANGELOG.md#4.0.0')}`;
                    this.e.log.nl();
                    this.e.log.info(`Looks like ${chalk_1.default.green(pkg)} isn't installed in this project.\n` + requiredMsg);
                    const installed = yield this.promptToInstallPkg({ pkg, saveDev: true });
                    if (!installed) {
                        throw new errors_1.FatalException(`${chalk_1.default.green(pkg)} is required for Ionic Lab to work properly.\n` + requiredMsg);
                    }
                    yield this.runLabServer(appUrl, labDetails);
                }
            }
            return labDetails;
        });
    }
    runLabServer(url, details) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pkg = yield this.e.project.requirePackageJson();
            const appName = this.e.project.config.get('name');
            const labArgs = [url, '--host', details.address, '--port', String(details.port)];
            const nameArgs = appName ? ['--app-name', appName] : [];
            const versionArgs = pkg.version ? ['--app-version', pkg.version] : [];
            if (details.ssl) {
                labArgs.push('--ssl', '--ssl-key', details.ssl.key, '--ssl-cert', details.ssl.cert);
            }
            const p = this.e.shell.spawn('ionic-lab', [...labArgs, ...nameArgs, ...versionArgs], { cwd: this.e.project.directory });
            this.emit('cli-utility-spawn', p);
            return new Promise((resolve, reject) => {
                p.on('error', err => {
                    reject(err);
                });
                process_1.onBeforeExit(() => tslib_1.__awaiter(this, void 0, void 0, function* () { return p.kill(); }));
                const log = this.e.log.clone();
                log.setFormatter(cli_framework_1.createPrefixedFormatter(chalk_1.default.dim('[lab]')));
                const ws = log.createWriteStream(cli_framework_1.LOGGER_LEVELS.INFO);
                const stdoutFilter = through2((chunk, enc, callback) => {
                    const str = chunk.toString();
                    if (str.includes('Ionic Lab running')) {
                        resolve();
                    }
                    else {
                        // no stdout
                    }
                    callback();
                });
                p.stdout.pipe(split2()).pipe(stdoutFilter).pipe(ws);
                p.stderr.pipe(split2()).pipe(ws);
            });
        });
    }
    promptToInstallPkg(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { pkgManagerArgs } = yield Promise.resolve().then(() => require('./utils/npm'));
            const [manager, ...managerArgs] = yield pkgManagerArgs(this.e.config.get('npmClient'), Object.assign({ command: 'install' }, options));
            const confirm = yield this.e.prompt({
                name: 'confirm',
                message: `Install ${chalk_1.default.green(options.pkg)}?`,
                type: 'confirm',
            });
            if (!confirm) {
                this.e.log.warn(`Not installing--here's how to install manually: ${chalk_1.default.green(`${manager} ${managerArgs.join(' ')}`)}`);
                return false;
            }
            yield this.e.shell.run(manager, managerArgs, { cwd: this.e.project.directory });
            return true;
        });
    }
    selectExternalIP(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let availableInterfaces = [];
            let chosenIP = options.address;
            if (options.address === exports.BIND_ALL_ADDRESS) {
                availableInterfaces = network_1.getExternalIPv4Interfaces();
                if (availableInterfaces.length === 0) {
                    if (options.externalAddressRequired) {
                        throw new errors_1.FatalException(`No external network interfaces detected. In order to use livereload with run/emulate you will need one.\n` +
                            `Are you connected to a local network?\n`);
                    }
                }
                else if (availableInterfaces.length === 1) {
                    chosenIP = availableInterfaces[0].address;
                }
                else if (availableInterfaces.length > 1) {
                    if (options.externalAddressRequired) {
                        this.e.log.warn('Multiple network interfaces detected!\n' +
                            'You will be prompted to select an external-facing IP for the livereload server that your device or emulator has access to.\n\n' +
                            `You may also use the ${chalk_1.default.green('--address')} option to skip this prompt.`);
                        const promptedIp = yield this.e.prompt({
                            type: 'list',
                            name: 'promptedIp',
                            message: 'Please select which IP to use:',
                            choices: availableInterfaces.map(i => ({
                                name: `${i.address} ${chalk_1.default.dim(`(${i.device})`)}`,
                                value: i.address,
                            })),
                        });
                        chosenIP = promptedIp;
                    }
                }
            }
            return [chosenIP, availableInterfaces];
        });
    }
}
exports.ServeRunner = ServeRunner;
class ServeBeforeHook extends hooks_1.Hook {
    constructor() {
        super(...arguments);
        this.name = 'serve:before';
    }
}
class ServeAfterHook extends hooks_1.Hook {
    constructor() {
        super(...arguments);
        this.name = 'serve:after';
    }
}
function serve(deps, inputs, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            const runner = yield deps.project.requireServeRunner();
            runner.on('cli-utility-spawn', cp => {
                cp.on('close', code => {
                    deps.log.nl();
                    deps.log.error(`A utility CLI has unexpectedly closed.\n` +
                        'The Ionic CLI will exit. Please check any output above for error details.');
                    process_1.processExit(1); // tslint:disable-line:no-floating-promises
                });
            });
            if (deps.project.name) {
                options['project'] = deps.project.name;
            }
            const opts = runner.createOptionsFromCommandLine(inputs, options);
            const details = yield runner.run(opts);
            return details;
        }
        catch (e) {
            if (e instanceof errors_1.RunnerException) {
                throw new errors_1.FatalException(e.message);
            }
            throw e;
        }
    });
}
exports.serve = serve;
